src/viewer/NativePdfViewer.jsx

```diff
@@ -0,0 +1,146 @@
+  },
+  print: {
+    enabled: true
+  },
+  openWith: {
+    enabled: true
+  },
+  bookmarks: {
+    enabled: true
+  },
+  search: {
+    enabled: false
+  },
+  autoClose: {
+    onPause: false
+  }
+}
+
+class NativePdfViewer extends Component {
+  state = {
+    opening: true,
+    error: null
+  }
+
+  nativeFileUrl = null
+
+  onOpen = () => {
+    this.setState({ opening: false })
+  }
+
+  onClose = () => {
+    this.props.onClose()
+  }
+
+  onError = e => {
+    console.warn(e)
+    this.setState({ error: e })
+  }
+
+  componentWillMount = async () => {
+    try {
+      if (this.props.file.isAvailableOffline) {
+        const file = await getNativeFile(this.props.file)
+        this.nativeFileUrl = file.nativeURL
+      } else {
+        const response = await cozy.client.files.downloadById(
+          this.props.file.id
+        )
+        const blob = await response.blob()
+        const file = await temporarySave(blob, this.props.file.name)
+        this.nativeFileUrl = file.nativeURL
+      }
+
+      this.openPdfInExternalViewer()
+    } catch (e) {
+      this.onError(e)
+    }
+  }
+
+  openPdfInExternalViewer = () => {
+    cordova.plugins.SitewaertsDocumentViewer.viewDocument(
+      this.nativeFileUrl,
+      'application/pdf',
+      {
+        ...VIEWER_OPTIONS,
+        documentView: {
+          closeLabel: this.props.t('Viewer.close')
+        },
+        navigationView: {
+          closeLabel: this.props.t('Viewer.close')
+        },
+        title: this.props.file.name
+      },
+      this.onOpen,
+      this.onClose,
+      this.openPdfWithSystem,
+      this.onError
+    )
+  }
+
+  openPdfWithSystem = async () => {
+    try {
+      document.addEventListener('resume', this.onSystemPdfViewerExit, false)
+      await openFileWithCordova(this.nativeFileUrl, 'application/pdf')
+      this.onOpen()
+    } catch (e) {
+      this.onError(e)
+    }
+  }
+
+  onSystemPdfViewerExit = () => {
+    document.removeEventListener('resume', this.onSystemPdfViewerExit)
+    this.onClose()
+  }
+
+  componentWillUnmount() {
+    document.removeEventListener('resume', this.onSystemPdfViewerExit)
+  }
+
+  render() {
+    const { file, t } = this.props
+    const { opening, error } = this.state
+
+    if (opening && !error)
+      return <Spinner size="xxlarge" middle="true" noMargin color="white" />
+    else if (error) return <NoViewer file={file} />
+    else
+      return (
+        <div
+          className={classNames(
+            styles['pho-viewer-noviewer'],
+            styles[`pho-viewer-noviewer--${file.class}`]
+          )}
+        >
+          <p className={styles['pho-viewer-filename']}>{file.name}</p>
+          <Button
+            theme="regular"
+            className={styles['pho-viewer-noviewer-download']}
+            onClick={this.openPdfInExternalViewer}
+          >
+            {t('Viewer.reopen')}
+          </Button>
+        </div>
+      )
+  }
+}
+
+export default translate()(NativePdfViewer)
```

Comment:
146 lignes pour un composant qui ne fait rien que le passe plat entre une UI et un plugin, c'est trop de responsabilités.

Reply:
Le nombre de lignes me semble un mauvais indicateur de ce genre de choses. Je vais bouger la configuration du plugin, et la gestion des fichiers offline sera généralisée à tous les viewers dans une future évolution.

Reply:
Le nombre de lignes est un bon indicateur du trop grand nombre de responsabilités d'une fonction ou d'une classe comme dit dans _Clean Code_ ou dans cet article https://medium.com/oril/10-tips-to-keep-your-code-more-clean-2fa9aafea1cf#0c65

Plus c'est court, moins y'a de responsibilités, moins y'a de paramètres, plus y'a de souplesse, de maintenabilité et de possibilité de réutilisabilité.